Add controllers

MaxiTrue opened the pull request:
Создан каркас приложения. Написана логика под классы User и Item

MaxiTrue committed:
feat:
1) Реализовать каркас приложения.
2) Создать классы основных сущностей User, Item, Booking, ItemRequest.
3) Создать под сущности User, Item классы контроллеры, сервисы, хранилища, мапперы, дто.
4) Реализовать хранение сущностей User, Item в памяти
5) Реализовать систему валидации приходящий объектов в запросах и систему exceptions

MaxiTrue committed:
feat:
1) Реализовать каркас приложения.
2) Создать классы основных сущностей User, Item, Booking, ItemRequest.
3) Создать под сущности User, Item классы контроллеры, сервисы, хранилища, мапперы, дто.
4) Реализовать хранение сущностей User, Item в памяти
5) Реализовать систему валидации приходящий объектов в запросах и систему exceptions

MaxiTrue committed:
fix: Изменил имя класса BookingController на BookController, ругался что конфликтуют бины с одинаковыми названиями, при реализации логики класса выясню почему

MaxiTrue committed:
fix: Исправил зависимости в pom файле

MaxiTrue committed:
fix: Правка стиля кода

MaxiTrue committed:
fix: Правка стиля кода

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/item/mapper/ItemMapper.java:
```diff
@@ -0,0 +1,77 @@
+    private final ItemStorage itemStorage;
+
+    public Item toEntityItem(ItemDto itemDto, long userId) throws Throwable {
+        User user = userStorage.getById(userId)
```
Мне кажется что тут объект-маппер занимается немного не своими обязанностями. Обязанность маппера - преобразовывать один объект в другой, а вот запросы в хранилище больше подходят для объекта-сервиса.
Но то что маппер сделан в виде спрингового бина - это правильно :)

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/user/storage/InMemoryUserStorage.java:
```diff
@@ -0,0 +1,61 @@
+    public Optional<User> getByEmail(String email) {
+        List<User> filterUsers = users.values().stream()
+                .filter(user1 -> user1.getEmail().equals(email))
+                .collect(Collectors.toList());
```
Если мы предполагаем что эмейл уникален (а судя по коду так и есть), тот тут можно сделать проще   :) 
return users.values().stream()
   .filter(user1 -> user1.getEmail().equals(email))
   .findFirst();

MaxiTrue replied:
Исправлено

eugene-lenkevich commented on src/main/java/ru/practicum/shareit/user/service/UserService.java:
```diff
@@ -0,0 +1,90 @@
+    private void validUserDto(String method, UserDto userDto) throws
+            ValidException, ObjectNotFoundException, NotUniqueEmailException {
+
+        if (!method.equals("POST")) {
```
В общем случае сервис не должен знать из какой среды его вызывают, в данном случае есть четкая зависимость от web. Причем можно заметить что код не универсален и зависит от конкретного кейса использования - от места вызова :)

MaxiTrue replied:
Исправлено

eugene-lenkevich reviewed (changes requested):
Привет, Максим, спасибо за PR

В целом все супер :)
Есть совсем небольшие замечания, посмотри комментарии ниже, пожалуйста

MaxiTrue committed:
refactor: Исправить архитектуру приложения:
1) Убрать вызовы слоя хранения из классов мапиров
2) Убрать из методов валидации проверку и привязку к эндпоинту вызова
refactor: В классе InMemoryUserStorage в методе getByEmail убрать получение коллекции в стриме и дальнейшую проверку на количсетво, реализовать сразу получение первого

MaxiTrue committed:
fix: Удалить неиспользуемые импорты

eugene-lenkevich reviewed (approved):
Все супер, спасибо!
